### DISCUSSION TITLE
How to use `mpremote.transport_serial.SerialTransport` programmatically?

### DISCUSSION THREAD

**Discussion by SandyWalsh** [micropython/micropython#15789] [General]
:wave: Since `pyboard.py` is deprecated, I'm trying to use `mpremote` as a python library to communicate with micropython on an ESP-8266 (nodemcu). 

```
# python3
Python 3.8.10 (default, Jul 29 2024, 17:02:10) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.

>>> import mpremote
>>> mpremote.__version__
'1.23.0'

>>> from mpremote import transport_serial as ts

>>> x = ts.SerialTransport("/dev/ttyUSB0")

>>> x.fs_listdir()   # Freezes until CTRL-C
^C
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.8/dist-packages/mpremote/transport_serial.py", line 321, in fs_listdir
    self.exec(cmd, data_consumer=repr_consumer)
  File "/usr/local/lib/python3.8/dist-packages/mpremote/transport_serial.py", line 285, in exec
    ret, ret_err = self.exec_raw(command, data_consumer=data_consumer)
  File "/usr/local/lib/python3.8/dist-packages/mpremote/transport_serial.py", line 271, in exec_raw
    self.exec_raw_no_follow(command)
  File "/usr/local/lib/python3.8/dist-packages/mpremote/transport_serial.py", line 236, in exec_raw_no_follow
    data = self.read_until(1, b">")
  File "/usr/local/lib/python3.8/dist-packages/mpremote/transport_serial.py", line 118, in read_until
    data = self.serial.read(min_num_bytes)
  File "/usr/local/lib/python3.8/dist-packages/serial/serialposix.py", line 565, in read
    ready, _, _ = select.select([self.fd, self.pipe_abort_read_r], [], [], timeout.time_left())
KeyboardInterrupt
```

If I attempt this from the cmdline 
```
$ mpremote connect /dev/ttyUSB0 fs ls
ls :
         228 boot.py
```
things work fine ... but driving it programmatically doesn't seem to work. 

I should also add the firmware version ... 
```
$ mpremote connect /dev/ttyUSB0
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] or Ctrl-x to exit this shell

>>> import machine
>>> machine.soft_reset()
MPY: soft reboot
MicroPython v1.23.0 on 2024-06-02; ESP module with ESP8266
Type "help()" for more information.
```

Am I just going about this wrong?

**Comment by SandyWalsh** [comment]
Ah, the :bulb: went on I think. 

```
>>> from mpremote import transport_serial as ts
>>> x = ts.SerialTransport("/dev/ttyUSB0")
>>> x.enter_raw_repl()
>>> x.fs_listdir()
[dir_result(name='base.py', st_mode=32768, st_ino=0, st_size=273), dir_result(name='boot.py', st_mode=32768, st_ino=0, st_size=228)]
>>> x.exit_raw_repl()
>>> 
```

**Reply by andrewleech** [reply]
Nice one, that's a clean example you arrived at there.

Just a heads up afaik that class isn't really locked down as a stable API for programmatic use yet, at least I think some refactoring was planned / done as part of https://github.com/micropython/micropython/pull/11777

I also wouldn't say pyboard.py is officially deprecated as such, it's still used to run the official test suite here, however certainly mpremote is the preferred cli tool. mpremote has the advantage of being pip installable too!

**Reply by shariltumin** [reply]
Yes, nice! Thanks for sharing. I'm interested in how you use mpremote in a script. Some examples would be nice.

**Reply by SandyWalsh** [reply]
Thanks @andrewleech ... noted. I figured best to hop on the train while I can :) I'll expect some breakage. 

@shariltumin Thanks, I'm using an esp8266 to drive gpio on an STM32 for some hardware-in-the-loop testing. I'm just starting to put a wrapper on this but it's starting to look something like this:

```
#!/usr/bin/env python3
from mpremote import transport_serial as ts

LED = 16

class GPIO:
    def __init__(self):
        self.serial = ts.SerialTransport("/dev/nucleo-a-gpio")
        self.serial.enter_raw_repl()

    def close(self):
        self.serial.exit_raw_repl()
        self.serial.close()

    def exec(self, cmd):
        self.serial.exec_raw_no_follow(cmd)

    def blink(self, times=10, delay_on=.2, delay_off=.2):
        # NOTE: cmds must have no leading whitespace or it will fail silently. 
        cmd = f"""
from machine import Pin, Signal
import time
led=Pin({LED}, Pin.OUT)
sig=Signal(led, invert=True)
for _ in range({times}):
    sig.value(1)
    time.sleep({delay_on})
    sig.value(0)
    time.sleep({delay_off})
"""
        self.exec(cmd)


class BoardUnderTestGPIO(GPIO):
    # Put board specific gpio mapping here
    pass


if __name__ == "__main__":
    gpio = GPIO()
    gpio.blink()
    gpio.close()
```

I still have to figure out how/when to expect responses from the `exec_raw_no_follow` call. That is, when to do a `follow()` call afterwards. 

I notice that much of `mpremote` assumes console output so it has a ways to go before it's truly an API. But it does a good job moving between raw repl and full repl.

**Reply by glenn20** [reply]
You *may* be interested in how I have wrapped up the `SerialTransport` interface in my `mpremote-path` tool. Most of the `SerialTransport` handling is done in [`board.py`](https://github.com/glenn20/mpremote-path/blob/main/src/mpremote_path/board.py).
